providers/implementations/encode_decode/decode_der2key.c

```diff
@@ -277,35 +277,52 @@ static int der2key_decode(void *vctx, OSSL_CORE_BIO *cin, int selection,
 
     ok = 0;                      /* Assume that we fail */
 
+    ERR_set_mark();
     if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0) {
         derp = der;
         if (ctx->desc->d2i_PKCS8 != NULL) {
             key = ctx->desc->d2i_PKCS8(NULL, &derp, der_len, ctx,
                                        pw_cb, pw_cbarg);
-            if (ctx->flag_fatal)
+            if (ctx->flag_fatal) {
+                ERR_clear_last_mark();
                 goto end;
+            }
         } else if (ctx->desc->d2i_private_key != NULL) {
             key = ctx->desc->d2i_private_key(NULL, &derp, der_len);
         }
-        if (key == NULL && orig_selection != 0)
+        if (key == NULL && orig_selection != 0) {
+            ERR_clear_last_mark();
             goto next;
+        }
     }
     if (key == NULL && (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
+        ERR_pop_to_mark();
+        ERR_set_mark();
         derp = der;
         if (ctx->desc->d2i_PUBKEY != NULL)
             key = ctx->desc->d2i_PUBKEY(NULL, &derp, der_len);
         else
             key = ctx->desc->d2i_public_key(NULL, &derp, der_len);
-        if (key == NULL && orig_selection != 0)
+        if (key == NULL && orig_selection != 0) {
+            ERR_clear_last_mark();
             goto next;
+        }
     }
     if (key == NULL && (selection & OSSL_KEYMGMT_SELECT_ALL_PARAMETERS) != 0) {
+        ERR_pop_to_mark();
+        ERR_set_mark();
         derp = der;
         if (ctx->desc->d2i_key_params != NULL)
             key = ctx->desc->d2i_key_params(NULL, &derp, der_len);
-        if (key == NULL && orig_selection != 0)
+        if (key == NULL && orig_selection != 0) {
+            ERR_clear_last_mark();
             goto next;
+        }
     }
+    if (key == NULL)
+        ERR_clear_last_mark();
+    else
+        ERR_pop_to_mark();
```

Comment:
Considering that this is done generically in `decoder_process`, I don't think you're on the right path here...

Reply:
The last of the
```
        ERR_pop_to_mark();
        ERR_set_mark();
```
inserted above prevents
```
asn1_check_tlen:crypto/asn1/tasn_dec.c:1156:CMP error: wrong tag:
asn1_item_embed_d2i:crypto/asn1/tasn_dec.c:322:CMP error: nested asn1 error:Type=EC_PRIVATEKEY
```
If you have a better suggestion how to prevent them please let me know.
Yet I suspect the intermittent errors must be purged there because
```
    if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0) {
        derp = der;
        if (ctx->desc->d2i_PKCS8 != NULL) {
            key = ctx->desc->d2i_PKCS8(NULL, &derp, der_len, ctx,
                                       pw_cb, pw_cbarg);
            if (ctx->flag_fatal) {
                ERR_clear_last_mark();
                goto end;
            }
        } else if (ctx->desc->d2i_private_key != NULL) {
            key = ctx->desc->d2i_private_key(NULL, &derp, der_len);
        }
        if (key == NULL && orig_selection != 0) {
            ERR_clear_last_mark();
            goto next;
        }
    }
    if (key == NULL && (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
        ERR_pop_to_mark();
        ERR_set_mark();
        derp = der;
        if (ctx->desc->d2i_PUBKEY != NULL)
            key = ctx->desc->d2i_PUBKEY(NULL, &derp, der_len);
        else
            key = ctx->desc->d2i_public_key(NULL, &derp, der_len);
        if (key == NULL && orig_selection != 0) {
            ERR_clear_last_mark();
            goto next;
        }
    }
    if (key == NULL && (selection & OSSL_KEYMGMT_SELECT_ALL_PARAMETERS) != 0) {
        ERR_pop_to_mark();
        ERR_set_mark();
        derp = der;
        if (ctx->desc->d2i_key_params != NULL)
            key = ctx->desc->d2i_key_params(NULL, &derp, der_len);
        if (key == NULL && orig_selection != 0) {
            ERR_clear_last_mark();
            goto next;
        }
    }
```
is a series of three tries to decode stuff, so if either of the first two fails and the third one succeeds the errors they produce need to be eliminated.

Reply:
So I doubt that this can be done generically in `decoder_process`.

Reply:
The bottom line here is that you got this:

```
ossl_store_handle_load_result:crypto/store/store_result.c:151:error: unsupported:
```

The rest of the errors are part of this...  they are errors that led to the situation where this unsupported error is reported.
So, the solution *should* be that when the unsupported error is correctly dealt with, with an `ERR_set_mark()` / `ERR_pop_to_mark()` in the right spot (I'm guessing somewhere in store_result.c), then the ASN1 error reports will go away at the same time.

Reply:
> The rest of the errors are part of this... they are errors that led to the situation where this unsupported error is reported.

This analysis cannot be fully adequate because
```
ossl_store_handle_load_result:crypto/store/store_result.c:151:CMP error: unsupported:
```
gets thrown before
```
asn1_check_tlen:crypto/asn1/tasn_dec.c:1156:CMP error: wrong tag:
asn1_item_embed_d2i:crypto/asn1/tasn_dec.c:322:CMP error: nested asn1 error:Type=EC_PRIVATEKEY
```

Reply:
Here is the stack trace of the 'unsupported' error - maybe this helps?
```
	ossl_store_handle_load_result() at store_result.c:151 0x7ffff7d177f5	
	file_load_construct() at file_store.c:409 0x7ffff7dbf691	
	decoder_process() at decoder_lib.c:604 0x7ffff7c1430c	
	pem2der_decode() at decode_pem2der.c:252 0x7ffff7d8a590	
	decoder_process() at decoder_lib.c:796 0x7ffff7c14759	
	OSSL_DECODER_from_bio() at decoder_lib.c:64 0x7ffff7c130ad	
	file_load_file() at file_store.c:535 0x7ffff7dbfa7f	
	file_load() at file_store.c:695 0x7ffff7dc0013	
	OSSL_STORE_load() at store_lib.c:413 0x7ffff7d145b3	
	load_key_certs_crls() at apps.c:975 0x48864d	
	load_key() at apps.c:577 0x48752f	
	load_key_pwd() at cmp.c:630 0x42bc91	
	setup_request_ctx() at cmp.c:1,553 0x42f5e4	
	setup_client_ctx() at cmp.c:1,891 0x4311c2	
	cmp_main() at cmp.c:2,777 0x433a99	
	do_cmd() at openssl.c:414 0x44daa0	
	main() at openssl.c:295 0x44d6cf	
```

Reply:
For the subsequent `nested asn1 error:Type=EC_PRIVATEKEY` it is
```
	asn1_item_embed_d2i() at tasn_dec.c:459 0x7ffff7b17f3d	
	ASN1_item_ex_d2i() at tasn_dec.c:129 0x7ffff7b16f75	
	ASN1_item_d2i() at tasn_dec.c:114 0x7ffff7b16ebe	
	d2i_EC_PRIVATEKEY() at ec_asn1.c:176 0x7ffff7bc874e	
	d2i_ECPrivateKey() at ec_asn1.c:932 0x7ffff7bca8ee	
	der2key_decode() at decode_der2key.c:291 0x7ffff7d87edd	
	decoder_process() at decoder_lib.c:796 0x7ffff7c14759	
	pem2der_decode() at decode_pem2der.c:252 0x7ffff7d8a590	
	decoder_process() at decoder_lib.c:796 0x7ffff7c14759	
	OSSL_DECODER_from_bio() at decoder_lib.c:64 0x7ffff7c130ad	
	file_load_file() at file_store.c:535 0x7ffff7dbfa7f	
	file_load() at file_store.c:695 0x7ffff7dc0013	
	OSSL_STORE_load() at store_lib.c:413 0x7ffff7d145b3	
	load_key_certs_crls() at apps.c:975 0x48864d	
	load_key() at apps.c:577 0x48752f	
	load_key_pwd() at cmp.c:630 0x42bc91	
	setup_request_ctx() at cmp.c:1,553 0x42f5e4	
	setup_client_ctx() at cmp.c:1,891 0x4311c2	
	cmp_main() at cmp.c:2,777 0x433a99	
	do_cmd() at openssl.c:414 0x44daa0	
	main() at openssl.c:295 0x44d6cf	
```